### PR TITLE
list/remove stage files from meta api

### PR DESCRIPTION
**Pull request by junnplus**
Signed-off-by: Ye Sijun <junnplus@gmail.com>

I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

list/remove stage files from meta api.

Add a `StageFile` struct for storage stage file:
```
pub struct StageFile {
    pub path: String,
    pub size: u64,
    pub md5: Option<String>,
    pub last_modified: DateTime<Utc>,
    pub creator: Option<UserIdentity>,
}
```

## Changelog

- Improvement

## Related Issues

Fixes #issue

### CODE REVIEW

File: common/meta/types/src/user_stage.rs [hunk 1]
```diff
@@ -227,18 +229,29 @@ pub struct UserStageInfo {
...
+                // It's internal, so we should prefix with stage name.
+                format!("/stage/{}/", self.stage_name)
+            }
         }
     }
 }
+
+#[derive(Default, Clone)]
+pub struct StageFile {
+    pub name: String,
```
Comment (Xuanwo) [inline common/meta/types/src/user_stage.rs:252]: How about using `path` to make the behavior clear?

File: common/meta/types/src/user_stage.rs [hunk 2]
```diff
@@ -227,18 +229,29 @@ pub struct UserStageInfo {
...
+    pub size: u64,
+    pub md5: String,
```
Comment (Xuanwo) [inline common/meta/types/src/user_stage.rs:254]: The file may lack `md5` sometimes:

- On underlying storage like `fs`, `hdfs`, and so on.
- Large objects (over 5GB) on s3, azblob, and so on.

Should we take this into consideration?

File: query/src/storages/system/stages_table.rs [hunk 3]
```diff
@@ -49,13 +49,38 @@ impl AsyncSystemTable for StagesTable {
         let mut copy_options: Vec<Vec<u8>> = Vec::with_capacity(stages.len());
         let mut file_format_options: Vec<Vec<u8>> = Vec::with_capacity(stages.len());
         let mut comment: Vec<Vec<u8>> = Vec::with_capacity(stages.len());
+        let mut number_of_files: Vec<Vec<u8>> = Vec::with_capacity(stages.len());
```
Comment (Xuanwo) [inline query/src/storages/system/stages_table.rs:52]: It's hard to calculate the number of files for external stages, what value should the external stage use?

File: query/src/sql/statements/statement_create_user_stage.rs [hunk 4]
```diff
@@ -61,6 +62,7 @@ impl AnalyzableStatement for DfCreateUserStage {
                     &self.location,
                     &self.credential_options,
                     &self.encryption_options,
+                    &self.endpoint_url,
```
Comment (Xuanwo) [inline query/src/sql/statements/statement_create_user_stage.rs:65]: Here is a consideration of accepting the user's endpoint.

Users could input endpoints like `127.0.0.1` to try to access our internal infra. We need to detect and avoid those things.

File: common/meta/types/src/user_stage.rs (hunk 2 above)
Comment (junnplus) [reply common/meta/types/src/user_stage.rs:254]: It should be an optional.

File: common/meta/types/src/user_stage.rs (hunk 1 above)
Comment (junnplus) [reply common/meta/types/src/user_stage.rs:252]: SGTM.

File: query/src/storages/system/stages_table.rs (hunk 3 above)
Comment (junnplus) [reply query/src/storages/system/stages_table.rs:52]: `number_of_file`s should be optional, it is a `None/NULL` for external stages.

**Comment by drmingdrmer** [comment]
It seems like it breaks the compatibility...

**Comment by junnplus** [comment]
> It seems like it breaks the compatibility...

What is compatible with?

**Comment by junnplus** [comment]
> The failure is emitted by the case `tests/suites/0_stateless/05_ddl/*stage_v2.*`.

I think you are referred to https://github.com/datafuselabs/databend/runs/6831188211?check_suite_focus=true , i have fixed it.

File: common/management/src/stage/stage_mgr.rs [hunk 5]
```diff
@@ -110,19 +123,170 @@ impl StageApi for StageMgr {
...
-        if res.prev.is_some() && res.result.is_none() {
-            Ok(())
-        } else {
-            Err(ErrorCode::UnknownStage(format!("Unknown stage {}", name)))
+    async fn drop_stage(&self, name: &str) -> Result<()> {
+        let mut retry = 0;
+        while retry < TXN_MAX_RETRY_TIMES {
+            retry += 1;
+
+            let stage_key = format!("{}/{}", self.stage_prefix, escape_for_key(name)?);
```
Comment (lichuang) [inline common/management/src/stage/stage_mgr.rs:131]: stage_key could be init outside the loop?

File: common/management/src/stage/stage_mgr.rs [hunk 6]
```diff
@@ -110,19 +123,170 @@ impl StageApi for StageMgr {
...
+            let file_key_prefix = format!("{}/{}/", self.stage_file_prefix, escape_for_key(name)?);
```
Comment (lichuang) [inline common/management/src/stage/stage_mgr.rs:132]: same as file_key_prefix

File: common/management/src/stage/stage_mgr.rs [hunk 7]
```diff
@@ -110,19 +123,170 @@ impl StageApi for StageMgr {
...
+        while retry < TXN_MAX_RETRY_TIMES {
+            retry += 1;
+
+            let stage_key = format!("{}/{}", self.stage_prefix, escape_for_key(name)?);
+            let file_key = format!(
+                "{}/{}/{}",
+                self.stage_file_prefix,
+                escape_for_key(name)?,
+                escape_for_key(&file.path)?
+            );
```
Comment (lichuang) [inline common/management/src/stage/stage_mgr.rs:176]: same as before

**Review by drmingdrmer — APPROVED** [review]

**Pull request merged**